server/handlers/validation.go

```diff
@@ -125,34 +77,150 @@ func validateUpdate(gun string, updates []storage.MetaUpdate, store storage.Meta
 	if _, ok := roles[targetsRole]; ok {
 		if t, err = validateTargets(targetsRole, roles, kdb); err != nil {
 			logrus.Error("ErrBadTargets: ", err.Error())
-			return ErrBadTargets{msg: err.Error()}
+			return nil, ErrBadTargets{msg: err.Error()}
 		}
 		repo.SetTargets(targetsRole, t)
 	}
 	logrus.Debug("Successfully validated targets")
 
-	var oldSnap *data.SignedSnapshot
-	oldSnapJSON, err := store.GetCurrent(gun, snapshotRole)
-	if _, ok := err.(*storage.ErrNotFound); err != nil && !ok {
-		// problem with storage. No expectation we can
-		// write if we can't read so bail.
-		logrus.Error("error reading previous snapshot: ", err.Error())
-		return err
-	} else if err == nil {
-		oldSnap = &data.SignedSnapshot{}
-		if err := json.Unmarshal(oldSnapJSON, oldSnap); err != nil {
-			oldSnap = nil
+	if _, ok := roles[snapshotRole]; ok {
+		var oldSnap *data.SignedSnapshot
+		oldSnapJSON, err := store.GetCurrent(gun, snapshotRole)
+		if _, ok := err.(*storage.ErrNotFound); err != nil && !ok {
+			// problem with storage. No expectation we can
+			// write if we can't read so bail.
+			logrus.Error("error reading previous snapshot: ", err.Error())
+			return nil, err
+		} else if err == nil {
+			oldSnap = &data.SignedSnapshot{}
+			if err := json.Unmarshal(oldSnapJSON, oldSnap); err != nil {
+				oldSnap = nil
+			}
+		}
+
+		if err := validateSnapshot(snapshotRole, oldSnap, roles[snapshotRole], roles, kdb); err != nil {
+			logrus.Error("ErrBadSnapshot: ", err.Error())
+			return nil, ErrBadSnapshot{msg: err.Error()}
+		}
+		logrus.Debug("Successfully validated snapshot")
+	} else {
+		// Check:
+		//	 - we have a snapshot key
+		//   - it matches a snapshot key signed into the root.json
+		// Then:
+		//   - generate a new snapshot
+		//   - add it to the updates
+		//
+		// TODO: Assumptions have been made about what is loaded into the repo.
```

Comment:
The root seems to have to be loaded, right?  Since if `SetRoot` above fails, this function will bail?

Also, is a `snapshots.json` that with no entry for targets metadata valid?  Previously, the hierarchy valid only checked for the presence of the snapshots file, and not the update file.  Maybe no targets metadata could just mean there are no targets yet?